File: ltms/go/setrule.go
// Set constraint rules for CWA.
// Converted from setrule.lisp.
package ltms

// LoadSetRules installs the rules for set membership reasoning.
func LoadSetRules(ltre *LTRE) {
	if ltre == nil {
		ltre = CurrentLTRE
	}
	old := CurrentLTRE
	CurrentLTRE = ltre
	defer func() { CurrentLTRE = old }()

	// Rule 1: When (set ?name) is true, install sub-rules for membership
	MakeRule([]interface{}{"SET", "?name"}, "TRUE",
		func(bindings map[string]interface{}, triggerNode *TmsNode) {
			name := bindings["?name"]

			// Sub-rule: When (?name MEMBERS ?construal1) is interned
			MakeRule([]interface{}{name, "MEMBERS", "?construal1"}, "INTERN",
				func(bindings2 map[string]interface{}, node2 *TmsNode) {
					construal1 := bindings2["?construal1"]

					// Sub-sub-rule: When (?name HAS-MEMBER ?new) is interned
					// and ?new is not in ?construal1
					MakeRule([]interface{}{name, "HAS-MEMBER", "?new"}, "INTERN",
						func(bindings3 map[string]interface{}, node3 *TmsNode) {
							newMember := bindings3["?new"]
							if construal1 == nil {
								return
							}
							// Test: new not in construal1
							if members, ok := construal1.([]interface{}); ok {
								for _, m := range members {
									if EqualForms(m, newMember) {
										return // already in set
									}
								}
							}
							// Get the datum nodes for the assertion
							setDatum := Referent([]interface{}{"SET", name}, false, CurrentLTRE)
							membersDatum := Referent([]interface{}{name, "MEMBERS", construal1}, false, CurrentLTRE)
							newDatum := Referent([]interface{}{name, "HAS-MEMBER", newMember}, false, CurrentLTRE)
							if setDatum != nil && membersDatum != nil && newDatum != nil {
								RAssert([]interface{}{":IMPLIES",
									[]interface{}{":AND",
										setDatum.TmsNode,
										membersDatum.TmsNode},
									[]interface{}{":NOT", newDatum.TmsNode}},
									":NOT-IN-SET")
							}
						})

					// Sub-sub-rule: Construal uniqueness
					MakeRule([]interface{}{name, "MEMBERS", "?construal2"}, "INTERN",
						func(bindings4 map[string]interface{}, node4 *TmsNode) {
							construal2 := bindings4["?construal2"]
							// Test: construal2 != construal1 and different sets
							if EqualForms(construal1, construal2) {
								return
							}
							// Check form ordering
							d1 := Referent([]interface{}{name, "MEMBERS", construal1}, false, CurrentLTRE)
							d2 := Referent([]interface{}{name, "MEMBERS", construal2}, false, CurrentLTRE)
							if d1 == nil || d2 == nil || d1.Counter >= d2.Counter {
								return
							}
							// Check symmetric difference
							if !hasSymmetricDifference(construal1, construal2) {
								return
							}
							setDatum := Referent([]interface{}{"SET", name}, false, CurrentLTRE)
							if setDatum != nil && d1 != nil && d2 != nil {
								RAssert([]interface{}{":NOT",
									[]interface{}{":AND",
										setDatum.TmsNode,
										d1.TmsNode,
										d2.TmsNode}},
									":CONSTRUAL-UNIQUENESS")
							}
						})
				})
		})

	// Rule 2: CWA-JUSTIFICATION rule
	MakeRule([]interface{}{"CWA-JUSTIFICATION", "?ante", "?conse"}, "INTERN",
		func(bindings map[string]interface{}, node *TmsNode) {
			ante := bindings["?ante"]
			conse := bindings["?conse"]
			cwajDatum := Referent([]interface{}{"CWA-JUSTIFICATION", ante, conse}, false, CurrentLTRE)
			if cwajDatum != nil {
				RAssert([]interface{}{":IMPLIES",
					cwajDatum.TmsNode,
					[]interface{}{":IMPLIES", ante, conse}},
					":CWA-JUSTIFICATION")
			}
		})
}

func hasSymmetricDifference(a, b interface{}) bool {
	aList, aOk := a.([]interface{})
	bList, bOk := b.([]interface{})
	if !aOk || !bOk {
		return !EqualForms(a, b)
	}
	// Check if there's any element in a not in b, or vice versa
	for _, ae := range aList {
		found := false
		for _, be := range bList {
			if EqualForms(ae, be) {
				found = true
				break
			}
		}
		if !found {
			return true
		}
	}
	for _, be := range bList {
		found := false
		for _, ae := range aList {
			if EqualForms(ae, be) {
				found = true
				break
			}
		}
		if !found {
			return true
		}
	}
	return false
}

File: ltms/java/SetRule.java
package ltms;

import java.util.*;

public class SetRule {

    public static void loadSetRules(LTRE ltre) {
        if (ltre == null) ltre = LTRE.currentLTRE;
        LTRE old = LTRE.currentLTRE;
        LTRE.currentLTRE = ltre;
        try {
            // Rule 1: When (SET ?name) is true, install sub-rules
            LtreRules.makeRule(Arrays.asList("SET", "?name"), "TRUE",
                (bindings, triggerNode) -> {
                    Object name = bindings.get("?name");

                    // Sub-rule: When (?name MEMBERS ?construal1) is interned
                    LtreRules.makeRule(Arrays.asList(name, "MEMBERS", "?construal1"), "INTERN",
                        (bindings2, node2) -> {
                            Object construal1 = bindings2.get("?construal1");

                            // Sub-sub-rule: (?name HAS-MEMBER ?new) not in construal1
                            LtreRules.makeRule(Arrays.asList(name, "HAS-MEMBER", "?new"), "INTERN",
                                (bindings3, node3) -> {
                                    Object newMember = bindings3.get("?new");
                                    if (construal1 == null) return;
                                    // Check not already in set
                                    if (construal1 instanceof List) {
                                        for (Object m : (List<?>) construal1) {
                                            if (LTRE.equalForms(m, newMember)) return;
                                        }
                                    }
                                    Datum setDatum = LTRE.referent(Arrays.asList("SET", name), false, LTRE.currentLTRE);
                                    Datum membersDatum = LTRE.referent(Arrays.asList(name, "MEMBERS", construal1), false, LTRE.currentLTRE);
                                    Datum newDatum = LTRE.referent(Arrays.asList(name, "HAS-MEMBER", newMember), false, LTRE.currentLTRE);
                                    if (setDatum != null && membersDatum != null && newDatum != null) {
                                        LTRE.assertFact(Arrays.asList(":IMPLIES",
                                            Arrays.asList(":AND", setDatum.tmsNode, membersDatum.tmsNode),
                                            Arrays.asList(":NOT", newDatum.tmsNode)),
                                            ":NOT-IN-SET", LTRE.currentLTRE);
                                    }
                                });

                            // Sub-sub-rule: Construal uniqueness
                            LtreRules.makeRule(Arrays.asList(name, "MEMBERS", "?construal2"), "INTERN",
                                (bindings4, node4) -> {
                                    Object construal2 = bindings4.get("?construal2");
                                    if (LTRE.equalForms(construal1, construal2)) return;
                                    Datum d1 = LTRE.referent(Arrays.asList(name, "MEMBERS", construal1), false, LTRE.currentLTRE);
                                    Datum d2 = LTRE.referent(Arrays.asList(name, "MEMBERS", construal2), false, LTRE.currentLTRE);
                                    if (d1 == null || d2 == null || d1.counter >= d2.counter) return;
                                    if (!hasSymmetricDifference(construal1, construal2)) return;
                                    Datum setDatum = LTRE.referent(Arrays.asList("SET", name), false, LTRE.currentLTRE);
                                    if (setDatum != null) {
                                        LTRE.assertFact(Arrays.asList(":NOT",
                                            Arrays.asList(":AND", setDatum.tmsNode, d1.tmsNode, d2.tmsNode)),
                                            ":CONSTRUAL-UNIQUENESS", LTRE.currentLTRE);
                                    }
                                });
                        });
                });

            // Rule 2: CWA-JUSTIFICATION rule
            LtreRules.makeRule(Arrays.asList("CWA-JUSTIFICATION", "?ante", "?conse"), "INTERN",
                (bindings, node) -> {
                    Object ante = bindings.get("?ante");
                    Object conse = bindings.get("?conse");
                    Datum cwajDatum = LTRE.referent(Arrays.asList("CWA-JUSTIFICATION", ante, conse), false, LTRE.currentLTRE);
                    if (cwajDatum != null) {
                        LTRE.assertFact(Arrays.asList(":IMPLIES",
                            cwajDatum.tmsNode,
                            Arrays.asList(":IMPLIES", ante, conse)),
                            ":CWA-JUSTIFICATION", LTRE.currentLTRE);
                    }
                });
        } finally {
            LTRE.currentLTRE = old;
        }
    }

    @SuppressWarnings("unchecked")
    static boolean hasSymmetricDifference(Object a, Object b) {
        if (!(a instanceof List) || !(b instanceof List)) return !LTRE.equalForms(a, b);
        List<Object> aList = (List<Object>) a;
        List<Object> bList = (List<Object>) b;
        for (Object ae : aList) {
            boolean found = false;
            for (Object be : bList) { if (LTRE.equalForms(ae, be)) { found = true; break; } }
            if (!found) return true;
        }
        for (Object be : bList) {
            boolean found = false;
            for (Object ae : aList) { if (LTRE.equalForms(ae, be)) { found = true; break; } }
            if (!found) return true;
        }
        return false;
    }
}
